### PR TITLE
Add Jasmine+Karma testing framework

### PR DESCRIPTION
**Pull request by danielbreves**
:v:

screen shot 2016-07-29 at 5 52 49 pm

/cc @zendesk/vegemite

### CODE REVIEW

File: karma.conf.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    // enable / disable colors in the output (reporters and logs)
+    colors: true,
+
+    // level of logging
+    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
+    logLevel: config.LOG_INFO,
+
+    // enable / disable watching file and executing tests whenever any file changes
+    autoWatch: true,
```
Comment (svizzari) [inline karma.conf.js]: Are singleRun and autoWatch mutually exclusive?

File: dist/assets/iframe.html [hunk 2]
```diff
@@ -8,8 +8,11 @@
   </head>
   <body>
     <section data-main></section>
+
+    <!-- CRUFT: if you change these remember to update karma.conf.js:17-18 -->
     <script src="https://cdn.jsdelivr.net/g/lodash@2.4.2(lodash.underscore.min.js),handlebarsjs@1.3.0,jquery@2.2.4,momentjs@2.9.0,bootstrap@2.3.2"></script>
```
Comment (svizzari) [inline dist/assets/iframe.html]: Could store externals config in a hash somewhere and inject url into compiled html via webpack. Then could give karma same hash/config so there was single source of truth?

**Comment by svizzari** [comment]
Pretty cool!

File: karma.conf.js (hunk 1 above)
Comment (danielbreves) [inline karma.conf.js]: Mmm yea, I guess so, not sure when you'd want this to be false...

Comment (svizzari) [inline karma.conf.js]: If you weren't a proponent for TDD, and didn't want an additional server/process running?

Comment (danielbreves) [inline karma.conf.js]: In that case, you would set `singleRun: true`, no? When `singleRun` is `true` `autoWatch` doesn't seem to make any difference.

Comment (svizzari) [inline karma.conf.js]: So, why is `singleRun` set to `true` here if that isn't the intention (default)?

**Comment by danielbreves** [comment]
@zendesk/vegemite this is ready for review!

**Comment by svizzari** [comment]
👍🏻

File: .eslintrc [hunk 3]
```diff
@@ -8,7 +8,8 @@
   "env": {
     "browser": true,
     "commonjs": true,
-    "es6": true
+    "es6": true,
+    "jasmine": true
```
Comment (sandlerr) [inline .eslintrc]: Would it be better to put another .eslintrc in the spec directory?

File: README.md [hunk 4]
```diff
@@ -43,7 +43,13 @@ or run the individual commands from the Procfile in separate terminals.
 Note: if you installed the Ruby version of foreman, you'll need to use `foreman start`.
 
 ## Testing
-WIP
+The App Scaffold is currently setup for testing with [Jasmine](http://jasmine.github.io/) (testing framework) and [Karma](https://karma-runner.github.io) (test runner). To run specs, run
+
+```
+karma start
```
Comment (sandlerr) [inline README.md]: do you have to globally install karma for that to work?

File: karma.conf.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    // start these browsers
+    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
+    browsers: ['PhantomJS'],
+
+    customLaunchers: {
+      // CRUFT: needed to load zaf_sdk with crossorigin=anonymous
```
Comment (sandlerr) [inline karma.conf.js]: what would it take to get rid of the cruft?

Comment (danielbreves) [inline karma.conf.js]: nginx would have to send CORS headers when serving zaf_sdk.js I believe.

File: README.md (hunk 4 above)
Comment (danielbreves) [inline README.md]: Yes, added to setup instructions on https://github.com/zendesk/app_scaffold/pull/17/commits/77ba887e660db6e6ea5263a24811d3c54681ad9e

File: spec/.eslintrc [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
+  "env": {
+    "jasmine": true,
+        "es6": false
```
Comment (svizzari) [inline spec/.eslintrc]: oh, the specs also contain es6, no?

File: karma.conf.js (hunk 5 above)
Comment (sandlerr) [inline karma.conf.js]: you mean from assets.zendesk.com? That's a CDN. I guess we can resolve this later?

File: spec/.eslintrc (hunk 6 above)
Comment (sandlerr) [inline spec/.eslintrc]: This should be inherited from the parent directory

File: README.md [hunk 7]
```diff
@@ -51,6 +51,9 @@ karma start
 
 Specs live under the `spec` directory and can be configured by editing the `karma.conf.js` file.
 
+## External Dependencies
+External dependencies are defined in the  [`webpack.config.js`](https://github.com/zendesk/app_scaffold/blob/master/webpack.config.js) as `externalAssets`. This ensures these dependencies are included on your app's `index.html` as well as the test suite.
```
Comment (svizzari) [inline README.md]: > as well as __in__ the test suite

?

File: spec/helpers/zaf.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
+function appendParams(url, params) {
+  let hashIndex = url.indexOf('#'),
+      urlBeforeHash = hashIndex > 0 ? url.substring(0, hashIndex) : url,
+      hasParams = urlBeforeHash.indexOf('?') > 0,
+      divider = hasParams > 0 ? '&' : '?',
```
Comment (sandlerr) [inline spec/helpers/zaf.js]: could you remove the `> 0` since hasParams is a boolean?

File: spec/helpers/zaf.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+function appendParams(url, params) {
```
Comment (sandlerr) [inline spec/helpers/zaf.js]: I'd have implemented this by passing in `window.location.search` splitting into an array on `&` and rejoining once new params are added, then you can just append `window.location.hash` on the end and `${window.location.origin}${window.location.pathname}` on the front. I'm not sure which would be more readable, my suggestion would probably be more verbose.

**Comment by sandlerr** [comment]
any thoughts how this will work once this is spun out to a node module?

File: spec/.eslintrc (hunk 6 above)
Comment (svizzari) [inline spec/.eslintrc]: Are you suggesting that's how it _should_ work, or how it currently does? If the latter, I don't think so, this local config would extend, and overwrite, the higher level config

Comment (danielbreves) [inline spec/.eslintrc]: This was a bad commit, I was testing that it would be extended, and it was.

**Comment by sandlerr** [comment]
:+1:

**Pull request merged**